javalib/src/main/scala/java/util/regex/Matcher.scala

```diff
@@ -275,5 +283,358 @@ object Matcher {
+    def setMatch(stringForGroupNumber: Int => String): Unit = {
+      this.matched = stringForGroupNumber(newGroup)
+      nodeType.fold(
+          r => (),
+          children => children.foreach(_.setMatch(stringForGroupNumber))
+      )
+    }
+
+    // Propagates the start position of this node to its descendants
+    def propagateStart(): Unit = {
+      nodeType.fold(
+        regex => (),
+        children => children.foldLeft(start) {
+          case (newStart, child) => child.setStartReturnEnd(newStart)
+        }
+      )
+    }
+
+    // Propagates the end position of this node to its descendants
+    def propagateEnd(): Unit = {
+      nodeType.fold(
+        regex => (),
+        children => children.foldRight(end) {
+          case (child, newEnd) => child.setEndReturnStart(newEnd)
+        }
+      )
+    }
+
+    def setEndReturnStart(newEnd: Int): Int = {
+      start = originalGroup match {
+        case OriginallyWrapped.Absolute() => newEnd
+        case _ =>
+          if (matched == null) -1
+          else newEnd - matched.length
+      }
+      propagateStart()
+      start
+    }
+    def setStartReturnEnd(offset: Int): Int = {
+      start = if (matched == null) -1
+              else offset
+
+      originalGroup match {
+        case OriginallyWrapped.Repeater(_) => propagateEnd()
+        case _ =>                          propagateStart()
+      }
+      originalGroup match {
+        case OriginallyWrapped.Absolute() => offset
+        case _ => end
+      }
+    }
+
+    def end: Int = start + (if (matched == null) 0 else matched.length)
+
+    def buildRegex: String =
+      originalGroup.fold(groupNum => "(", (left: String, right: String) => "((?:" + left) +
+      nodeType.     fold(regex => regex,
+        children => "(?:" + children.map(_.buildRegex).mkString("") + ")") +
+      originalGroup.fold(groupNum => ")", (left: String, right: String) => ")" + right + ")")
+
+    def obtainMapping: Map[Int, Node] = {
+       originalGroup.fold(groupNum => Map(groupNum -> this), (_, _) => Map[Int, Node]()) ++
+       nodeType.     fold(regex => Map[Int, Node](), children =>
+         children.foldLeft(Map[Int, Node]()){
+           case (mapping, child) => mapping ++ child.obtainMapping
+         })
+    }
+
+    def simplify: Node = {
+      this match {
+        case Node(OriginallyGroupped(nextGroupIndex),
+                  ParentNode(Seq(Node(OriginallyWrapped("", ""),
+                                      RegexLeaf(regex))))) =>
+          Node(OriginallyGroupped(nextGroupIndex), RegexLeaf(regex))
+        case _ => this
+      }
+
+    }
+  }
+
+  // Wraps every char in the regexp in a group to find the intermediate positions.
+  // Same for subgroups.
+  // (?:AB)C => (?:(AB))(?:(C))
+  // Transform:  (A(B))C(D)  to  (((A)(B))(C)(D)) so that using the length
+  // of the 1st group and the 3rd, we can obtain the start position of D.
+  private[regex] def parseRegex(pattern: String): Node = {
```

Comment:
The two steps are done at the same time for performance reason. You could build the original regex using the (now deleted on your request) `buildOriginalRegex` function. So it's indeed a parse.
But the parse is done so that there is enough information (original group, repeaters, children...) to build the new regular expression from another kind of traversal of the node.